### PR TITLE
Record canvas api calls in DB

### PR DESCRIPTION
**Pull request by MueezKhan246**
Description:
Changes to canvas client to record canvas api calls in DB.

JIRA:
https://2u-internal.atlassian.net/jira/software/c/projects/ENT/issues/ENT-8275

### CODE REVIEW

File: integrated_channels/canvas/client.py [hunk 1]
```diff
@@ -632,7 +693,18 @@ def _get_canvas_user_courses_by_id(self, user_id):
...
         path = f'/api/v1/users/{user_id}/courses'
         get_users_courses_url = urljoin(self.enterprise_configuration.canvas_base_url, path)
+        start_time = time.time()
         rsps = self.session.get(get_users_courses_url)
+        duration_seconds = time.time() - start_time
+        self.IntegratedChannelAPIRequestLogs.store_api_call(
+            enterprise_customer=self.enterprise_configuration.enterprise_customer,
+            enterprise_customer_configuration_id=self.enterprise_configuration.id,
+            endpoint=get_users_courses_url,
+            payload={},
```
Comment (hamzawaleed01) [inline integrated_channels/canvas/client.py:703]: should we set it a string?

File: integrated_channels/canvas/client.py [hunk 2]
```diff
@@ -665,7 +737,18 @@ def _handle_canvas_assignment_retrieval(
...
         # Check if the course assignment already exists
         canvas_assignments_url = CanvasUtil.course_assignments_endpoint(self.enterprise_configuration, course_id)
+        start_time = time.time()
         resp = self.session.get(canvas_assignments_url)
+        duration_seconds = time.time() - start_time
+        self.IntegratedChannelAPIRequestLogs.store_api_call(
+            enterprise_customer=self.enterprise_configuration.enterprise_customer,
+            enterprise_customer_configuration_id=self.enterprise_configuration.id,
+            endpoint=canvas_assignments_url,
+            payload={},
```
Comment (hamzawaleed01) [inline integrated_channels/canvas/client.py:747]: same as above

File: integrated_channels/canvas/client.py [hunk 3]
```diff
@@ -701,7 +784,19 @@ def _handle_canvas_assignment_retrieval(
...
                 next_page = CanvasUtil.determine_next_results_page(resp)
                 if next_page:
+                    start_time = time.time()
                     resp = self.session.get(next_page)
+                    duration_seconds = time.time() - start_time
+                    self.IntegratedChannelAPIRequestLogs.store_api_call(
+                        enterprise_customer=self.enterprise_configuration.enterprise_customer,
+                        enterprise_customer_configuration_id=self.enterprise_configuration.id,
+                        endpoint=next_page,
+                        payload={},
```
Comment (hamzawaleed01) [inline integrated_channels/canvas/client.py:794]: same as above

File: integrated_channels/canvas/client.py [hunk 4]
```diff
@@ -747,7 +853,18 @@ def _handle_canvas_assignment_submission(self, grade, course_id, assignment_id,
...
             }
         }
+        start_time = time.time()
         submission_response = self.session.put(submission_url, json=submission_data)
+        duration_seconds = time.time() - start_time
+        stringify_and_store_api_record(
+            enterprise_customer=self.enterprise_configuration.enterprise_customer,
+            enterprise_customer_configuration_id=self.enterprise_configuration.id,
+            endpoint=submission_url,
+            data=json.dumps(submission_data),
```
Comment (hamzawaleed01) [inline integrated_channels/canvas/client.py:863]: `stringify_and_store_api_record` is responsible for stringifying the `data`

File: integrated_channels/canvas/utils.py [hunk 5]
```diff
@@ -42,7 +43,18 @@ def find_root_canvas_account(enterprise_configuration, session):
...
         """
         url = urljoin(enterprise_configuration.canvas_base_url, "/api/v1/accounts")
+        start_time = time.time()
         resp = session.get(url)
+        duration_seconds = time.time() - start_time
+        integrated_channel_request_log_model().store_api_call(
+            enterprise_customer=enterprise_configuration.enterprise_customer,
+            enterprise_customer_configuration_id=enterprise_configuration.id,
+            endpoint=url,
+            payload={},
```
Comment (hamzawaleed01) [inline integrated_channels/canvas/utils.py:53]: same as above

File: integrated_channels/cornerstone/views.py [hunk 6]
```diff
@@ -163,16 +156,12 @@ def get(self, request, *args, **kwargs):
...
-            time_taken=duration_seconds,
-            status_code=200,
-            response_body=json.dumps(data)
+        # TODO remove following logs (temporarily added)
+        logger.info(
+            f"[Cornerstone]: request.headers={request.headers}"
+            f"GET params={request.GET}"
+            f"enterprise_config={enterprise_config}"
+            f"enterprise_config.id={enterprise_config.id}"
+            f"data={data}"
```
Comment (hamzawaleed01) [inline integrated_channels/cornerstone/views.py:165]: this can be removed now?

File: integrated_channels/integrated_channel/models.py [hunk 7]
```diff
@@ -926,12 +926,6 @@ def __str__(self):
             f", status_code: {self.status_code}"
         )
 
-    def __repr__(self):
-        """
-        Return uniquely identifying string representation.
-        """
-        return self.__str__()
-
```
Comment (hamzawaleed01) [inline integrated_channels/integrated_channel/models.py:934]: is it intentionally removed?

File: integrated_channels/utils.py [hunk 8]
```diff
@@ -501,3 +508,42 @@ def get_enterprise_client_by_channel_code(channel_code):
...
+        if not isinstance(data, str):
+            try:
+                # Check if data is a dictionary, list, or tuple then convert to JSON string
+                if isinstance(data, (dict, list, tuple)):
+                    data = json.dumps(data)
+                else:
+                    # If it's another type, simply convert to string
+                    data = str(data)
+            except (TypeError, ValueError) as e:
+                print(f"Error during stringification: {e}")
```
Comment (hamzawaleed01) [inline integrated_channels/utils.py:536]: can we use `generate_formatted_log` here?

File: integrated_channels/utils.py [hunk 9]
```diff
@@ -501,3 +508,42 @@ def get_enterprise_client_by_channel_code(channel_code):
...
+                enterprise_customer=enterprise_customer,
+                enterprise_customer_configuration_id=enterprise_customer_configuration_id,
+                endpoint=endpoint,
+                payload=data,
+                time_taken=time_taken,
+                status_code=status_code,
+                response_body=response_body,
+            )
+        except Exception as e:   # pylint: disable=broad-except
+            print(f"Failed to store data in the database: {e}")
```
Comment (hamzawaleed01) [inline integrated_channels/utils.py:549]: and here as well?

**Review by hamzawaleed01 — CHANGES_REQUESTED** [review]

File: integrated_channels/canvas/client.py (hunk 1 above)
Comment (MueezKhan246) [reply integrated_channels/canvas/client.py]: it should be empty string.
updated

File: integrated_channels/cornerstone/views.py (hunk 6 above)
Comment (MueezKhan246) [reply integrated_channels/cornerstone/views.py]: it can be.
updated.

File: integrated_channels/integrated_channel/models.py (hunk 7 above)
Comment (MueezKhan246) [reply integrated_channels/integrated_channel/models.py]: no it was not intentionally removed?
updated.

File: integrated_channels/canvas/client.py (hunk 4 above)
Comment (MueezKhan246) [reply integrated_channels/canvas/client.py]: yes, submission_data can be sent directly.

File: integrated_channels/canvas/client.py [hunk 10]
```diff
@@ -559,7 +598,18 @@ def _delete(self, url):
...
             url (str): The canvas url to send delete requests to.
         """
+        start_time = time.time()
         delete_response = self.session.delete(url, data='{"event":"conclude"}')
+        duration_seconds = time.time() - start_time
+        stringify_and_store_api_record(
+            enterprise_customer=self.enterprise_configuration.enterprise_customer,
+            enterprise_customer_configuration_id=self.enterprise_configuration.id,
+            endpoint=url,
+            data='{"event":"conclude"}',
```
Comment (hamzawaleed01) [inline integrated_channels/canvas/client.py:608]: we can create a variable for it and use on line 602 as well

File: integrated_channels/cornerstone/views.py [hunk 11]
```diff
@@ -163,16 +156,4 @@ def get(self, request, *args, **kwargs):
...
-        headers_json = json.dumps(headers_dict)
-        IntegratedChannelAPIRequestLogs.store_api_call(
-            enterprise_customer=enterprise_customer,
-            enterprise_customer_configuration_id=enterprise_config.id,
-            endpoint=request.get_full_path(),
-            payload=f"Request Headers: {headers_json}",
-            time_taken=duration_seconds,
-            status_code=200,
-            response_body=json.dumps(data)
-        )
```
Comment (hamzawaleed01) [inline integrated_channels/cornerstone/views.py:177]: 🤔

File: integrated_channels/canvas/client.py (hunk 10 above)
Comment (MueezKhan246) [reply integrated_channels/canvas/client.py]: yes a variable can be used
updated.

File: integrated_channels/cornerstone/views.py (hunk 11 above)
Comment (MueezKhan246) [reply integrated_channels/cornerstone/views.py]: while pulling changes i was only keeping the changes from your branch that were according to the scope of my PR
that is why this was getting removed
updated.

File: integrated_channels/cornerstone/views.py [hunk 12]
```diff
@@ -106,6 +106,7 @@ def get(self, request, *args, **kwargs):
         IntegratedChannelAPIRequestLogs = apps.get_model(
             "integrated_channel", "IntegratedChannelAPIRequestLogs"
         )
+        enterprise_customer_uuid = request.GET.get('ciid')
```
Comment (hamzawaleed01) [inline integrated_channels/cornerstone/views.py:109]: is this intentional? 🤔

Comment (MueezKhan246) [reply integrated_channels/cornerstone/views.py:109]: i checked out on master 
and it was missing there.

File: integrated_channels/canvas/client.py [hunk 13]
```diff
@@ -309,7 +314,18 @@ def cleanup_duplicate_assignment_records(self, courses):
...
+                duration_seconds = time.time() - start_time
+                self.IntegratedChannelAPIRequestLogs.store_api_call(
+                    enterprise_customer=self.enterprise_configuration.enterprise_customer,
+                    enterprise_customer_configuration_id=self.enterprise_configuration.id,
+                    endpoint=canvas_assignments_url,
+                    payload='',
+                    time_taken=duration_seconds,
+                    status_code=resp.status_code,
+                    response_body=resp.text,
+                )
```
Comment (sameenfatima78) [inline integrated_channels/canvas/client.py:328]: Question: Shouldn't we be using `stringify_and_store_api_record` here just like we are doing at all other places?

**Review by sameenfatima78 — COMMENTED** [review]
looks good, just one question.

**Review by hamzawaleed01 — APPROVED** [review]

**Review by sameenfatima78 — APPROVED** [review]

**Pull request merged**